### PR TITLE
Feat/soft delete

### LINKED ISSUE #79: Allow for soft cleanup (use deleted_at instead of deleting the row) for info for admins later on

### PR DESCRIPTION
**Pull request by lesagi**
Hi,

I added the option for soft for both InMemory and Relational, need to add Redis

I made it optional, and also I prevent getting the URL if the URL was soft deleted, please let me know what you think

I've merged the release branch to prevent conflicts in the future, that made the PR a bit hard to follow, my new Impl starts at:

"Merge remote-tracking branch 'origin/release/0.2.0' into release/0.2.0"

#79

### CODE REVIEW

File: src/services/storage/drivers/relational/index.ts [hunk 1]
```diff
@@ -69,15 +70,15 @@ export class RelationalStorage implements StorageDriver {
...
 
-		public async get(id: string, options = { withInfo: false }): Promise<StoredUrl | UrlWithInformation> {
+		public async get(id: string, options = { withInfo: false, isAuthorized: false }): Promise<StoredUrl | UrlWithInformation> {
 			let storedUrl, urlInfo
+			storedUrl = await this.storage.db
+				.table<StoredUrl & { serial?: number }>('urls')
+				.select('*')
+				.where('id', id)
+				.first()
+			if (storedUrl?.deletedAt && !options.isAuthorized) throw new UnauthorizedError();
```
Comment (SnirShechter) [inline src/services/storage/drivers/relational/index.ts:80]: If a user asks for a URL with id=4 (doesn't exist), it'll receive a 404. If they request a URL with id=3 (deleted), they'll receive an UnauthorizedError. This gives away the fact we use soft-delete instead of hard-delete (bad).

File: src/services/storage/drivers/relational/index.ts [hunk 2]
```diff
@@ -93,9 +94,14 @@ export class RelationalStorage implements StorageDriver {
...
-		public async delete(id: string): Promise<void> {
-			await this.storage.db.table<StoredUrl>('urls').where('id', id).delete()
-			return
+		public async delete(id: string, options: { hardDelete: false }): Promise<void> {
+			if (!await this.storage.db.table<StoredUrl>('urls').where('id', id)) throw new NotFoundError();
+
+			if (options.hardDelete) {
+				await this.storage.db.table<StoredUrl>('urls').where('id', id).delete()
+			} else {
+				await this.storage.db.table<StoredUrl>('urls').update({ deletedAt: new Date().toISOString() }).where('id', id)
```
Comment (SnirShechter) [inline src/services/storage/drivers/relational/index.ts:103]: I think it'll give the locale date (not GMT). Can you check?

File: src/services/storage/drivers/relational/index.ts [hunk 3]
```diff
@@ -93,9 +94,14 @@ export class RelationalStorage implements StorageDriver {
...
+			}
 		}
 
 		public async deleteOverdue(timespanMs: number): Promise<number> {
```
Comment (SnirShechter) [inline src/services/storage/drivers/relational/index.ts:107]: What about deleteOverdue? We need it to soft-delete too.

**Review by SnirShechter — CHANGES_REQUESTED** [review]

**Review by Danielg212 — COMMENTED** [review]
@lesagi please see Snir comments

File: src/services/storage/drivers/relational/index.ts (hunk 1 above)
Comment (lesagi) [reply src/services/storage/drivers/relational/index.ts]: Changed both to 404

File: src/services/storage/drivers/relational/index.ts [hunk 4]
```diff
@@ -69,15 +70,15 @@ export class RelationalStorage implements StorageDriver {
 	url = new (class RelationalUrlStorage {
 		constructor(public storage: RelationalStorage) { }
 
-		public async get(id: string, options = { withInfo: false }): Promise<StoredUrl | UrlWithInformation> {
+		public async get(id: string, options = { withInfo: false, isAuthorized: false }): Promise<StoredUrl | UrlWithInformation> {
```
[the remark this answers is not in the document]
Comment (lesagi) [reply src/services/storage/drivers/relational/index.ts]: Took the name you suggested

File: src/services/storage/drivers/relational/index.ts (hunk 3 above)
Comment (lesagi) [reply src/services/storage/drivers/relational/index.ts:106]: added a deleteOverdue + migration to support the soft delete in postgres

**Comment by lesagi** [comment]
> @lesagi please see Snir comments

Hi, Snir was updated yesterday, it was WIP, not can be reviewed again, thanks for the heads up!
